### ISSUE TITLE
Black box question: curl -sS https://getcomposer.org/installer | sudo php

### ISSUE THREAD

**Issue by dianakleban** [susanBuck/dwa15-fall2017#64]
I know this code should be a black box for us for now, but I would like to understand it a little bit more.
```$ curl -sS https://getcomposer.org/installer | sudo php```
I am not sure what are all parts of the command, but my understanding is that it is installing composer on the server? I looked at the composer documentation, but cannot find all parts of the command there.

**Comment by susanBuck** [comment]
In plain English, the translation of this command is *Download the Composer installer and then run it with PHP as the root user*.

[Full breakdown via ExplainShell...](https://explainshell.com/explain?cmd=curl+-sS+https%3A%2F%2Fgetcomposer.org%2Finstaller+%7C+sudo+php)

The above link covers the details, but the pipe character (`|`) explanation is verbose and not straightforward IMO. So here's my explanation:

The pipe character lets you "pipe" one command to another.

So in this case, it's saying take the results of the command on the left (`curl -sS https://getcomposer.org/installer`) and pipe it to the command on the right (`sudo php`).

If you have any follow-up questions from the ExplainShell output, let us know.

**Comment by ysrinu** [comment]
pipeline (|) operator is to pass output of one program (or utility) as input to another.
Here `curl -sS https://getcomposer.org/installer` is one program (curl with parameters) and it's output is fed as input to `sudo php`

Note that this is different than redirection operators (< or >) where one of the operations is file i/o
for example: `sort a.txt > b.txt `(sort file a.txt and write the output to file b.txt)

**Comment by gregorycoleman** [comment]
Maybe a simpler way to understand pipes is to think of them as temporary files.

Suppose  `$ curl -sS https://getcomposer.org/installer | sudo php` were written as two separate commands. The first command would download the installer script to a temporary file

`$ curl -sS https://getcomposer.org/installer > /tmp/tempfile`

And the second command would run it through php

`sudo php <  /tmp/tempfile`

**Comment by HeinrichTremblay** [comment]
I found a [well explained video](https://www.youtube.com/watch?v=7XUibDYw4mc) on youtube about `curl`. It is often used to download files from the command line or to make [api calls](http://docs.oracle.com/cloud/latest/marketingcs_gs/OMCAB/Developers/GettingStarted/API%20requests/curl-requests.htm).

**Comment by dianakleban** [comment]
Thank you for the explanation everyone. I just want to double check that when we download it, it is getting installed on the production server itself, correct?

**Comment by ysrinu** [comment]
It's installed wherever that curl command is executed. In this case if you are running on digital ocean (production env), then yes it's isntalled on production.

**Comment by TheLivingNick** [comment]
On top of the temp file concept, I find it useful to think of it as a function returning something, so 
`curl -sS https://getcomposer.org/installer | sudo php`
could be conceived as 
`sudo php(curl -sS https://getcomposer.org/installer)`
